compiler/qsc_doc_gen/src/generate_docs.rs

```diff
@@ -140,51 +158,110 @@ pub fn generate_docs(
     let compilation = Compilation::new(additional_sources, capabilities, language_features);
-    let mut files: Files = vec![];
+    let mut files: FilesWithMetadata = vec![];
 
     let display = &CodeDisplay {
         compilation: &compilation,
     };
 
     let mut toc: FxHashMap<Rc<str>, Vec<String>> = FxHashMap::default();
-    for (_, unit) in &compilation.package_store {
+
+    for (package_id, unit) in &compilation.package_store {
+        let is_current_package = compilation.current_package_id == Some(package_id);
+        let package_kind;
+        match package_id {
+            // Core package is always included in the compilation.
+            PackageId::CORE => package_kind = PackageKind::Core,
+
+            // Standard package is currently always included, but this isn't enforced by the compiler.
+            _ if package_id == 1.into() => package_kind = PackageKind::StandardLibrary,
+
+            // This package could be user code if current package is specified.
+            _ if is_current_package => {
+                package_kind = PackageKind::UserCode;
+            }
+
+            _ => {
+                if let Some(alias) = compilation.dependencies.get(&package_id) {
+                    // This is a direct dependency of the user code.
+                    package_kind = PackageKind::AliasedPackage(alias.to_string());
+                } else {
+                    // This is not a package user can access (an indirect dependency).
+                    continue;
+                }
+            }
+        }
+
         let package = &unit.package;
         for (_, item) in &package.items {
-            if let Some((ns, line)) = generate_doc_for_item(package, item, display, &mut files) {
+            if let Some((ns, line)) = generate_doc_for_item(
+                package,
+                package_kind.clone(),
+                is_current_package,
+                item,
+                display,
+                &mut files,
+            ) {
                 toc.entry(ns).or_default().push(line);
             }
         }
     }
 
-    generate_toc(&mut toc, &mut files);
+    // We want to sort documentation files in a meaningful way.
+    // First, we want to put files for the current project, if it exists
+    // Then we want to put explicit dependencies of the current project, if they exist
+    // Then we want to add built-in std package. And finally built-in core package.
+    // Namespaces within packages should be sorted alphabetically and
+    // items with a namespace should be also sorted alphabetically.
+    // Also, items without any metadata (table of content) should come last
+    files.sort_by_key(|file| {
+        (
+            file.0.package.clone(),
+            file.0.namespace.clone(),
+            file.0.name.clone(),
+        )
+    });
+
+    let mut result: Files = files
+        .into_iter()
+        .map(|(metadata, name, content)| (name, Arc::from(metadata.to_string().as_str()), content))
```

Comment:
Files and FilesWithMetadata are both maintained only in this file. In this conversion, I see the ordering of their fields differs. It would be good if they had the same ordering of fields, as that would make reasoning between the two a little more straightforward.

Reply:
Updated.